### PR TITLE
Added code to trigger nw initiated detach

### PR DESCRIPTION
**Pull request by vthiruveedula**

### CODE REVIEW

File: service/init.go [hunk 1]
```diff
@@ -534,3 +541,86 @@ func (amf *AMF) SendNFProfileUpdateToNrf() {
...
+	for _, supi := range slice.DeletedImsis {
+		amfSelf := context.AMF_Self()
+		ue, ok = amfSelf.AmfUeFindBySupi("imsi-" + supi)
+		if !ok {
+			logger.CfgLog.Infof("the UE [%v] is not Registered with the 5G-Core", supi)
+			continue
+		}
+		//publish the event to ue channel
+		configMsg := context.ConfigMsg{
+			Supi: supi,
```
Comment (UpadhyayaAnkur) [inline service/init.go:617]: Possible to include ue-context itself as 2 lookups happening in current flow via amfSelf.AmfUeFindBySupi()
1st called in HandleImsiDeleteFromNetworkSlice
2nd called in  UeConfigHandler()

File: gmm/sm.go [hunk 2]
```diff
@@ -442,14 +451,26 @@ func DeregisteredInitiated(state *fsm.State, event fsm.EventType, args fsm.ArgsT
...
+		SetDeregisteredState(amfUe, util.AnTypeToNas(accessType))
 		logger.GmmLog.Debugln(event)
 	case fsm.ExitEvent:
 		logger.GmmLog.Debugln(event)
 	default:
 		logger.GmmLog.Errorf("Unknown event [%+v]", event)
 	}
 }
+
+func SetDeregisteredState(amfUe *context.AmfUe, anType uint8) {
```
Comment (UpadhyayaAnkur) [inline gmm/sm.go:463]: Prefer as following(method on AmfUe object) 
func (amfUe *context.AmfUe) SetDeregisteredState(anType)

File: gmm/handler.go [hunk 3]
```diff
@@ -1530,6 +1530,56 @@ func AuthenticationProcedure(ue *context.AmfUe, accessType models.AccessType) (b
...
+	accessType models.AccessType) (err error) {
+
+	anType := util.AnTypeToNas(accessType)
+	if ue.CmConnect(accessType) {
+		//setting reregistration required flag to true
+		gmm_message.SendDeregistrationRequest(ue.RanUe[accessType], anType, true, 0)
+	} else {
+		SetDeregisteredState(ue, anType)
+	}
+	// TODO: Need to implement Nudm_SDM_Unsubscribe
```
Comment (UpadhyayaAnkur) [inline gmm/handler.go:1543]: Check if AMF is sending callback URL to UDM, only then AMF need to send Unsubscribe to UDM.

**Review by UpadhyayaAnkur — APPROVED** [review]

File: service/init.go [hunk 4]
```diff
@@ -534,3 +541,85 @@ func (amf *AMF) SendNFProfileUpdateToNrf() {
...
+		}
+		// TODO:more than one slice, trigger UEConfigurationUpdate Procedure
+		if nwInitieadDetach {
+			gmm.GmmFSM.SendEvent(ue.State[models.AccessType__3_GPP_ACCESS], gmm.NwInitiatedDeregistrationEvent, fsm.ArgsType{
+				gmm.ArgAmfUe:      ue,
+				gmm.ArgAccessType: models.AccessType__3_GPP_ACCESS,
+			})
+		}
+	} else {
+		logger.CfgLog.Warnf("Ue is not in Registered State")
```
Comment (UpadhyayaAnkur) [inline service/init.go:598]: probably we should add "nwInitieadDetach" handling for all states, like not registered condition here

File: service/init.go (hunk 1 above)
Comment (vthiruveedula) [reply service/init.go:673]: yes .. required in two places, i will optimise in my next commits

File: gmm/handler.go (hunk 3 above)
Comment (vthiruveedula) [reply gmm/handler.go:1549]: yes, its required .. created new bug for unsubscribe handling in AMF and UDM

**Comment by vthiruveedula** [comment]
retest container

**Pull request merged**